src/runtime/library_module.cc

```diff
@@ -123,20 +126,63 @@ void ImportModuleBlob(const char* mblob, std::vector<Module>* mlist) {
   dmlc::Stream* stream = &fs;
   uint64_t size;
   CHECK(stream->Read(&size));
+  std::vector<Module> modules;
+  bool has_import_tree = false;
   for (uint64_t i = 0; i < size; ++i) {
     std::string tkey;
     CHECK(stream->Read(&tkey));
-    if (tkey == "c") continue;
-    std::string fkey = "module.loadbinary_" + tkey;
-    const PackedFunc* f = Registry::Get(fkey);
-    CHECK(f != nullptr)
+    // Currently, _lib is for DSOModule, but we
+    // don't have loadbinary function for it currently
+    if (tkey == "_lib") {
+      auto dso_module = Module(make_object<LibraryModuleNode>(lib));
+      // allow lookup of symbol from dso root (so all symbols are visible).
+      if (auto *ctx_addr =
```

Comment:
NOTE: given that multiple libs can share the same symbol, we cannot repetitively patch the ctx_addr of each dso_module into it(they will get overriden). Instead, let us simply patch the root module's address into ctx_addr at the end of the loop

Reply:
This logic can be unified with the root_mod patching logic in the caller

Reply:
I did patch the addr of root_mod outsider in the caller simply, but if the root_mod is not dso module, it can not work, we must patch the ctx addr using the dso mod. What is the suggestion of it?

Reply:
If the root module is not the dso module, patching the root address should still work, because it only requiried for GetFuncFromEnv, and GetFuncFromEnv will recursively call GetFunction from the children

Reply:
Let us describe the example I have tested. I construct one gpu lib resnet50_gpu_lib. Then, I export the graph runtime module as the root module. That is to say, graph runtime contains llvm module, then llvm module contains cuda module. But if we patch the root address of graph runtime module, we will meet error like `fuse_conv2d_xxx` can not find.  

I find we call
```
  InitContextFunctions([lib](const char* fname) {
      return lib->GetSymbol(fname);
    });
```, which calls GetFuncFromEnv. But the lib is just for DSOModule. The problem is here?

Reply:
OK, we can change https://github.com/apache/incubator-tvm/blob/f823c577215d1b87191e598f6b1d3fef182452e6/src/runtime/module.cc#L118 to set query_imports=true